core/lib/zksync_core/src/commitment_generator/mod.rs

```diff
@@ -0,0 +1,230 @@
+        latency.observe();
+
+        Ok(AuxCommitments {
+            events_queue_commitment,
+            bootloader_initial_content_commitment,
+        })
+    }
+
+    async fn prepare_input(
+        connection: &mut StorageProcessor<'_>,
+        l1_batch_number: L1BatchNumber,
+    ) -> anyhow::Result<CommitmentInput> {
+        let header = connection
+            .blocks_dal()
+            .get_l1_batch_header(l1_batch_number)
+            .await?
+            .context("header is missing for batch")?;
+        let tree_data = connection
+            .blocks_dal()
+            .get_l1_batch_tree_data(l1_batch_number)
+            .await?
+            .context("tree data is missing for batch")?;
+
+        // TODO(PLA-731): ensure that the protocol version is always available.
+        let protocol_version = header
+            .protocol_version
+            .unwrap_or_else(ProtocolVersionId::last_potentially_undefined);
+        let common = CommitmentCommonInput {
+            l2_to_l1_logs: header.l2_to_l1_logs,
+            rollup_last_leaf_index: tree_data.rollup_last_leaf_index,
+            rollup_root_hash: tree_data.hash,
+            bootloader_code_hash: header.base_system_contracts_hashes.bootloader,
+            default_aa_code_hash: header.base_system_contracts_hashes.default_aa,
+            protocol_version,
+        };
+        let touched_slots = connection
+            .storage_logs_dal()
+            .get_touched_slots_for_l1_batch(l1_batch_number)
+            .await?;
+        let touched_hashed_keys: Vec<_> =
+            touched_slots.keys().map(|key| key.hashed_key()).collect();
+        let previous_values = connection
+            .storage_logs_dal()
+            .get_previous_storage_values(&touched_hashed_keys, l1_batch_number)
+            .await?;
+        let l1_batches_for_initial_writes = connection
+            .storage_logs_dal()
+            .get_l1_batches_and_indices_for_initial_writes(&touched_hashed_keys)
+            .await?;
+
+        let input = if protocol_version.is_pre_boojum() {
+            let mut initial_writes = Vec::new();
+            let mut repeated_writes = Vec::new();
+            for (key, value) in touched_slots.into_iter().sorted_by_key(|(key, _)| *key) {
+                let hashed_key = key.hashed_key();
+                if previous_values[&hashed_key].unwrap_or_default() != value {
+                    let (initial_write_l1_batch_number, index) =
+                        l1_batches_for_initial_writes[&hashed_key];
+                    if initial_write_l1_batch_number == l1_batch_number {
+                        initial_writes.push(InitialStorageWrite {
+                            index,
+                            key: key.hashed_key_u256(),
+                            value,
+                        })
+                    } else {
+                        repeated_writes.push(RepeatedStorageWrite { index, value })
```

Comment:
Maybe it makes sense to do a sanity check (or a condition if we're not sure it'll always hold) that `initial_write_l1_batch_number < l1_batch_number`? Ditto for post-Boojum batches.